observ/store.py

```diff
@@ -0,0 +1,135 @@
+import copy
+from functools import partial, wraps
+from inspect import signature
+from typing import Callable, TypeVar
+
+from .dep import Dep
+from .observables import (
+    reactive,
+    readonly,
+    shallow_reactive,
+    to_raw,
+)
+from .watcher import Watcher
+
+
+T = TypeVar("T", bound=Callable)
+
+
+registry = {}
+
+
+def mutation(fn: T) -> T:
+    registry[fn] = "mutation"
+
+    @wraps(fn)
+    def inner(self, *args, **kwargs):
+        return fn(self, *args, **kwargs)
+
+    return inner
+
+
+def computed(fn: T) -> T:
+    parameters = signature(fn).parameters
+    if len(parameters) == 1 and "self" in parameters:
+        registry[fn] = "computed"
+
+        @wraps(fn)
+        def inner(self):
+            return fn(self)
+
+        return inner
+    else:
+        watcher = Watcher(fn)
+
+        @wraps(fn)
+        def getter():
+            if watcher.dirty:
+                watcher.evaluate()
+            if Dep.stack:
+                watcher.depend()
+            return watcher.value
+
+        getter.__watcher__ = watcher
+        return getter
+
+
+registered_computed_props = {}
+
+
+class Store:
+    """
+    Store that tracks mutations to state in order to enable undo/redo functionality
+    """
+
+    def __init__(self, state):
+        self._present = reactive(state)
+        # Because we can't work with diff patches (yet) we'll have to copy
+        # over the whole state. When triggering undo/redo, we'll therefore
+        # trigger all the watchers / computed expressions...
+        self._past = shallow_reactive([copy.deepcopy(to_raw(state))])
+        self._future = shallow_reactive([])
+        self.state = readonly(state)
+
+        computed_props = registered_computed_props.setdefault(id(self), {})
```

Comment:
This can be a local dict on self right? No need for a global store or anything

Reply:
Just looked at the code and remembered why I did this. We override `__getattribute__` and if you get a local prop on `self`, you will end up in an endless loop. So that's why I stored it outside the object, in order to circumvent that. I can see if there is another way of dealing with the endless loop.

Reply:
Ok, I can just check for the name of the local dict.

Reply:
You can grab it of `super()` I think, or otherwise use `super().__getattribute__("name of your dict")` to avoid hitting your implementation of `__getattribute__`